pkg/apiserver/experiment/experiment.go

```diff
@@ -938,381 +159,437 @@ func (s *Service) listExperiments(c *gin.Context) {
 	var (
-		chaosKind *v1alpha1.ChaosKind
-		ok        bool
 		exp       *core.Experiment
+		expDetail *Detail
 	)
 
 	kubeCli, err := clientpool.ExtractTokenAndGetClient(c.Request.Header)
 	if err != nil {
-		_ = c.Error(utils.ErrInvalidRequest.WrapWithNoMessage(err))
+		u.SetAPIError(c, u.ErrBadRequest.WrapWithNoMessage(err))
+
 		return
 	}
 
 	uid := c.Param("uid")
 	if exp, err = s.archive.FindByUID(context.Background(), uid); err != nil {
 		if gorm.IsRecordNotFoundError(err) {
-			c.Status(http.StatusInternalServerError)
-			_ = c.Error(utils.ErrInvalidRequest.New("the experiment is not found"))
+			u.SetAPIError(c, u.ErrNotFound.New("Experiment "+uid+" not found"))
 		} else {
-			c.Status(http.StatusInternalServerError)
-			_ = c.Error(utils.ErrInternalServer.NewWithNoMessage())
+			u.SetAPIError(c, u.ErrInternalServer.WrapWithNoMessage(err))
 		}
+
 		return
 	}
 
-	kind := exp.Kind
-	ns := exp.Namespace
-	name := exp.Name
-	force := c.DefaultQuery("force", "false")
+	ns, name, kind := exp.Namespace, exp.Name, exp.Kind
 
-	ctx := context.TODO()
-	chaosKey := types.NamespacedName{Namespace: ns, Name: name}
+	if chaosKind, ok := v1alpha1.AllKinds()[kind]; ok {
+		expDetail = s.findChaosInCluster(c, kubeCli, types.NamespacedName{Namespace: ns, Name: name}, chaosKind.Chaos)
+
+		if expDetail == nil {
+			return
+		}
+	} else {
+		u.SetAPIError(c, u.ErrBadRequest.New("Kind "+kind+" is not supported"))
 
-	if chaosKind, ok = v1alpha1.AllKinds()[kind]; !ok {
-		c.Status(http.StatusBadRequest)
-		_ = c.Error(utils.ErrInvalidRequest.New(kind + " is not supported"))
 		return
 	}
-	if err := kubeCli.Get(ctx, chaosKey, chaosKind.Chaos); err != nil {
-		if apierrors.IsNotFound(err) {
-			c.Status(http.StatusNotFound)
-			_ = c.Error(utils.ErrNotFound.NewWithNoMessage())
-		} else {
-			c.Status(http.StatusInternalServerError)
-			_ = c.Error(utils.ErrInternalServer.WrapWithNoMessage(err))
-		}
+
+	c.JSON(http.StatusOK, expDetail)
+}
+
+func (s *Service) findChaosInCluster(c *gin.Context, kubeCli client.Client, namespacedName types.NamespacedName, chaos runtime.Object) *Detail {
+	if err := kubeCli.Get(context.Background(), namespacedName, chaos); err != nil {
+		u.SetAPImachineryError(c, err)
+
+		return nil
+	}
+
+	getChaosResult := reflect.ValueOf(chaos).MethodByName("GetChaos").Call(nil)[0].Elem()
+
+	return &Detail{
+		Experiment: Experiment{
+			ObjectBase: core.ObjectBase{
+				Namespace: reflect.ValueOf(chaos).Elem().FieldByName("Namespace").String(),
+				Name:      reflect.ValueOf(chaos).Elem().FieldByName("Name").String(),
+				Kind:      getChaosResult.FieldByName("Kind").String(),
+				UID:       getChaosResult.FieldByName("UID").String(),
+				Created:   getChaosResult.FieldByName("StartTime").Interface().(time.Time).Format(time.RFC3339),
+			},
+			Status: status.GetChaosStatus(chaos.(v1alpha1.InnerObject)),
+		},
+		KubeObject: core.KubeObjectDesc{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: v1alpha1.GroupVersion.String(),
```

Comment:
> And here are a lot of codes working with reflect, like:
> 
> ```go
> reflect.ValueOf(chaos).Elem().FieldByName("APIVersion").String()
> ```
> 
> we could use type assert with `metav1.Object` and `schema.ObjectKind`, then use getter or setter. And this PR has been enough huge, I think I could help for improving that in the next PRs.

@STRRL I currently hardcoded the `APIVersion` this way, but I don't know if that's a good idea since it's not dynamically fetched, but I guess even if it's not dynamic, there shouldn't be an error here? Because if we change the version in the future, `v1alpha1.GroupVersion` will be updated as well.

Reply:
let me dig into the kubernetes machinery to findout 🤔

Reply:
I consulted while solving this problem: https://github.com/kubernetes/kubernetes/pull/63972. Hope it helps.

Reply:
I think we could just use:

```go
			TypeMeta: metav1.TypeMeta{
				APIVersion: chaos.GetObjectKind().GroupVersionKind().GroupVersion().String(),
				Kind:       chaos.GetObjectKind().GroupVersionKind().Kind,
			},
```

Reply:
because `runtime.Object` contains the method:

```go
type Object interface {
	GetObjectKind() schema.ObjectKind
	DeepCopyObject() Object
}
```

and `json.Unmarshal` could resolve fields: `APIVersion` and `Kind` properly into `runtime.TypeMeta`

Reply:
I thought so too, but you can check my PR link above and the related issues, and while I didn't dig into the details, it looks like GVK is not serialized out in the k8s API because it would omit a lot of duplicate GVK fields?

Reply:
Another way is using https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/scheme to retrieve the GVK dynamically. But I think if we can use the existing method to do the GVK retrieving, we can do it without introducing new dependencies.

Reply:
I looked at #2217 and it looks like we can only use the second method to get the GVK currently after it is merged.

Reply:
> I thought so too, but you can check my PR link above and the related issues, and while I didn't dig into the details, it looks like GVK is not serialized out in the k8s API because it would omit a lot of duplicate GVK fields?

I find one related PR: https://github.com/chaos-mesh/chaos-mesh/pull/1368.

I only tried `chaos.GetObjectKind().GroupVersionKind().GroupVersion().String()` in the `archive.go`, which definitely have `APIVersion` and `Kind` in the serialized json.

Let me take a look at other files like `experiment.go` :D

Reply:
Ok. 👌🏽

Reply:
@g1eny0ung, in PR #1368, we use `apiutil.GVKForObject()` to fetch the GVK each time. 

I think we could also use `apiutil.GVKForObject()` as a temporary solution in this PR.

Reply:
> I only tried chaos.GetObjectKind().GroupVersionKind().GroupVersion().String() in the archive.go, which definitely have APIVersion and Kind in the serialized json.
> 
> Let me take a look at other files like experiment.go :D

fetching GVK with reflection DOES NOT WORK in `experiment.go` and `schedule.go`, because the GVK info could only referenced by `runtime.Schema` in kubernetes apimachinary.


> @g1eny0ung, in PR #1368, we use apiutil.GVKForObject() to fetch the GVK each time.
> 
> I think we could also use apiutil.GVKForObject() as a temporary solution in this PR.

Maybe we could make the `v1alpha1.AllKinds()` better: filling the `TypeMeta` of each Chaos when `register` the `ChaosKind`. But this operation needs a `runtime.Schema`, and we have not provide one in package `api/v1alpha1`. And I think provide `runtime.Schema` in `api/v1alpha1` is not good. :/

I will raise an issue about that for discussing the next step of refactoring.

Reply:
> I think we could also use apiutil.GVKForObject() as a temporary solution in this PR.

I think so too, and this could be a temporary solution. Looking forward to making this part easier in the future.

Reply:
Already done.